rust/datafusion/src/physical_plan/distinct_expressions.rs

```diff
@@ -0,0 +1,303 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Implementations for DISTINCT expressions, e.g. `COUNT(DISTINCT c)`
+
+use std::cell::RefCell;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use arrow::array::ArrayRef;
+use arrow::array::{
+    Int16Array, Int32Array, Int64Array, Int8Array, PrimitiveArrayOps, UInt16Array,
+    UInt32Array, UInt64Array, UInt8Array,
+};
+use arrow::datatypes::{DataType, Schema};
+use arrow::record_batch::RecordBatch;
+
+use fnv::FnvHashSet;
+
+use crate::error::{ExecutionError, Result};
+use crate::logical_plan::ScalarValue;
+use crate::physical_plan::expressions::Column;
+use crate::physical_plan::hash_aggregate::AggregateMode;
+use crate::physical_plan::{Accumulator, AggregateExpr, PhysicalExpr};
+
+/// Enumeration of types that can be accumulated into a distinct set of values.
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+enum DistinctScalarValue {
```

Comment:
These are used as the values in the iterator's `HashSet`, and unfortunately it's not possible to derive `Hash` for `ScalarValue` since it nests `f32` and `f64`. 

As you suggested, `GroupByScalar` is a good candidate here. (I assumed you meant `GroupByScalar` rather than `GroupByKey`?)